### ISSUE TITLE
Evaluate LMDB as a replacement for leveldb and/or bdb

### ISSUE THREAD

**Issue by bitcartel** [zcash/zcash#2194]
LMDB
https://symas.com/offerings/lightning-memory-mapped-database/
https://github.com/LMDB/lmdb

Benchmarks
https://symas.com/offerings/lightning-memory-mapped-database/benchmarks/
https://www.influxdata.com/benchmarking-leveldb-vs-rocksdb-vs-hyperleveldb-vs-lmdb-performance-for-influxdb/

Caveat: use fsync
http://www.openldap.org/lists/openldap-devel/201410/msg00008.html

**Comment by zookozcash** [comment]
The author of LMDB says that Zcash is a scam and that its authors are criminals: https://twitter.com/hyc_symas/status/818789330948919296

This doesn't _necessarily_ mean that LMDB is badly designed or badly implemented (;-)), but it might complicate things if we wanted to ask him for user support, submit bug reports, etc.

I am a huge fan of sqlite and would happily give my 👍 to using that. It has a long track record of reliability in many deployments, [their test policy](https://www.sqlite.org/testing.html) is top notch and we could learn from it! I have had extensive — and entirely positive — experience working with sqlite. The developer docs are great.

If anyone wants to argue against sqlite on performance grounds, that's fine, but I'm going to listen to this argument only if you bring a realistic performance test that shows (at least approximately) how many z-addresses/t-addresses/transactions/etc. incur how much of a performance penalty in a realistic scenario. My guess is that sqlite is more than performant enough for most or all practical scenarios.

**Comment by veikkoeeva** [comment]
A vote for SQLite here too, an excellent cross-platform and development friendly choice.

An added note, as this may not be evident, but if ZCash were to adopt the convention that all queries will be explicitly identified and used so that the ZCash "engine" would care only that the variable and column names and types remain the same, it would be possible to switch the DB engine or the implementation on the fly. This gives the added benefit one can adapt the structures taking account the hardware and amount of data, for instance, as in sending "computation and data" between the engine and the program. Prior art is availabe at https://github.com/dotnet/orleans/tree/master/src/OrleansSQLUtils, see different scripts (the queries are keyed [like this](https://github.com/dotnet/orleans/blob/master/src/OrleansSQLUtils/CreateOrleansTables_SqlServer.sql#L252) and loaded and used by these keys). This way one can make use of the various tooling suites and backends depending on any platform variables. This is, I suppose, mostly like Stackoverflow operates too. As to why this might be important for a running system, maybe [this post mortem](https://blogs.msdn.microsoft.com/bharry/2016/02/06/a-bit-more-on-the-feb-3-and-4-incidents/) indicates as to why. As the amount of data grows or there are problems, one could adjust queries on the fly.

On tangential, this would also allow one to transparently to make use features such as at https://www.pipelinedb.com/ (assuming one would switch from SQLite to PostgreSQL) or pipe data through the DB interface to any other storage (e.g. https://msdn.microsoft.com/en-us/library/mt143171.aspx) by simply modifying the queries appropriate &ndash; as it would be transparent to ZCash. I.e. making ZCash perhaps more readily integrable to various other systems.

**Comment by radix42** [comment]
One thing I'd like to make note of is that currently all calls to access either database (LevelDB or Berkeley DB) are very much peculiar to the database in use. I'd recommend first factoring out all calls that result in database access into some nice abstraction layer, starting of course with the current databases that are in use. Then safely swapping out the backend(s) will be a whole lot easier. I bloodied my head against this some months ago when I made the start of a switch to lmdb. It was not pretty because of this.

**Comment by veikkoeeva** [comment]
@radix42 I second that too. Which is, in fact, how the aforementioned _Storage Providers_ in the just linked Orleans work. This is repeating the obvious, but for completeness sake, there's an interface and an implementation class and then a loader the loads the implementation. One backend is ADO(.NET), i.e. "relational databases" and that specific backend works as just mentioned.

**Comment by hyc** [comment]
@bitcartel re: fsync, the bug those researchers noted was actually fixed years before they reported it
http://www.openldap.org/lists/openldap-devel/201501/msg00002.html

Nevertheless, we added a workaround to avoid the issue on older kernels:
https://github.com/LMDB/lmdb/commit/462dc097451834477b597447af69c5acc93182b7

There are no special considerations needed for building or using current versions of LMDB wrt fsync/fdatasync.

@zookozcash actually, LMDB's design and implementation are unmatched, as verified by multiple independent reviewers. There is no other DB engine in the world that can offer LMDB's efficiency, reliability, and transactional consistency all in under 64KB of object code. Nothing else even comes close.

It is also exhaustively documented inline, with doxygen. And, unlike some project leaders, I can stand up at a software conference and explain exactly how my software works and why it is more efficient than everything else out there.

**Comment by ebfull** [comment]
@nathan-at-least We tentatively believe that sqlite would be a nice replacement for Berkeley DB, and would like you feedback. If we don't hear from you soon, we will assume you agree.

**Comment by zookozcash** [comment]
@hyc: Thanks for the note. So, I take it you've changed your mind that we are criminals?

In any case, there are two different uses of DBs in Zcash — 1. the chain state DB that contains UTXOs, tree states, and stuff like, and 2. the wallet data file. The latter is where we currently use BerkeleyDB, and we definitely want to get rid of it (unreliable, problematic licensing, frequent backward-incompatible releases), and the latter doesn't ([I guess](https://github.com/zcash/zcash/issues/2194#issuecomment-288827202)) need better-than-sqlite performance.

I also with what @veikkoeeva and @radix42 are saying above — we should use an abstraction layer: https://github.com/zcash/zcash/issues/2194#issuecomment-289211140, https://github.com/zcash/zcash/issues/2194#issuecomment-289028249

**Comment by zookozcash** [comment]
Oh wait, I didn't notice/understand the "dig" against me at the end of [Howard's comment](https://github.com/zcash/zcash/issues/2194#issuecomment-289307827): "unlike some project leaders, I can stand up at a software conference and explain exactly how my software works". (For those who aren't in the know, some Monero folks — Howard among them — are fond of disparagingly pointing out that [I don't understand how zk-SNARKs work](https://www.reddit.com/r/Monero/comments/5yae1b/monero_is_a_sincere_attempt_at_privacy_and_i/depyur8/).)

I really don't think we should rely on LMDB, for the reason that it could cause problems if we needed to collaborate with the maintainer and he behaves like this. Also, because as far as I can tell we don't _need_ anything that LMDB is uniquely good for.

**Comment by zookozcash** [comment]
Conclusion: we're not going to use LMDB. Re-open this ticket if (a) we need functionality not supplied by sqlite or another widely-used tool with good testing, docs, and track-record, and (b) we think we can get LMDB support without working with Howard Chu, either by sourcing it from someone else or by doing it ourselves.

**Comment by anonimal** [comment]
>but it might complicate things if we wanted to ask him for user support, submit bug reports, etc.
>*...might...*

>I really don't think we should rely on LMDB, for the reason that it could cause problems if we needed to collaborate with the maintainer and he behaves like this.
>*...for the reason that it could cause problems...*
>*...for the reason...*
>*...could...*

>(b) we think we can get LMDB support without working with Howard Chu, either by sourcing it from someone else or by doing it ourselves.

Hi @zookozcash. What support would be needed that isn't already publicly provided? Did @hyc ever turn down support? Did he ever say he *would* turn down support?

I understand there are differences in opinion but I also think that LMDB is quite the impressive database.

**Comment by daira** [comment]
@anonimal: The choice of any dependency is not purely technical. If the maintainer of a potential dependency has repeatedly displayed contempt and made meritless accusations against a project, as @hyc has done for Zcash (on Twitter particularly), then they're effectively disqualifying their software from use by that project.

**Comment by P-E-Meunier** [comment]
Btw, I'm just seeing this thread now, but I implemented an LMDB-like key-value store about two years ago as part of the Pijul project. It uses similar algorithms, is written in 100% Rust, and has similar performance to LMDB. It has been fairly reliable for over a year now.

The reasons for reimplementation a transactional B tree from scratch included 1. lack of features in LMDB (I needed to clone databases in O(1)), and 2. relations with the author of LMDB (our project was judged "minor" rather than "criminal", but still…).

The project is called Sanakirja, it is available on https://nest.pijul.com/pijul_org/sanakirja.
- Proper bibliography shows that it is *not* the ultimate datastructure for databases, self-improving databases could be better on average, even though AFAIK, no one has written a transactional, on-disk version of this yet.
- My benchmarks show it compares favorably to LevelDB in the cases I've tried, but this doesn't say much about how it would behave in your specific case.

Since this project was prompted by a rejection of feature suggestion/collaboration, we're obviously as open as possible to suggestions, benchmarks, contribution, etc (Rust's code of conduct is the only restriction).
